common/compositor/gl/glprogram.cpp

```diff
@@ -126,8 +126,83 @@ static std::string GenerateFragmentShader(int layer_count) {
   return fragment_shader_stream.str();
 }
 
+static GLint LoadPreBuiltBinary(GLint gl_program,
+                                void *binary, long size) {
+  GLint status;
+
+  /* check if glProgramBinaryOES exists */
+  if (!glProgramBinaryOES)
+    return 0;
+
+  /* currently GL_MESA_program_binary_formats is not exposed by MESA drv
+   * TODO: will enable this once this is fixed in MESA
+   */
+
+  glProgramBinaryOES(gl_program, GL_PROGRAM_BINARY_FORMAT_MESA, binary,
+                     size);
+
+  glGetProgramiv(gl_program, GL_LINK_STATUS, &status);
+  if (status)
+      return gl_program;
+
+  return 0;
+}
+
 static GLint GenerateProgram(unsigned num_textures,
                              std::ostringstream *shader_log) {
+  GLint status;
+  GLint program = glCreateProgram();
+  void *binary_prog;
+  long binary_sz;
+
+  if (!program) {
+    if (shader_log)
+      *shader_log << "Failed to create program."
+                  << "\n";
+    return 0;
+  }
+
+  /* try to load prebuilt shader program from files */
+  std::ostringstream shader_program_fname;
+  shader_program_fname << PREBUILT_SHADER_FILE_PATH"/hwc_shader_prog_"
+                       << num_textures
+                       <<".shader_test.bin";
+
+  FILE *shader_prog_fp;
+
+  shader_prog_fp = fopen(shader_program_fname.str().c_str(), "rb");
+
+  if (shader_prog_fp) {
+    /* check the size of file */
+    fseek(shader_prog_fp, 0, SEEK_END);
+    binary_sz = ftell(shader_prog_fp);
+    rewind(shader_prog_fp);
+
+    char *binary_prog = (char *)malloc(binary_sz);
+    if (binary_prog) {
+      if (fread(binary_prog, 1, binary_sz, shader_prog_fp) == binary_sz) {
+        fclose(shader_prog_fp);
+        status = LoadPreBuiltBinary(program, binary_prog, binary_sz);
+        free(binary_prog);
+        if (status) {
+          if (shader_log)
+            *shader_log << "Pre-built shader program binary has been loaded "
+                        << "Successfully (from files)\n";
+          return program;
+        }
+      } else {
+          free(binary_prog);
+          fclose(shader_prog_fp);
+      }
+    } else {
+        fclose(shader_prog_fp);
```

Comment:
formatting here looks off. When in doubt, use clang-format.

Reply:
@strassek : yeah thanks for the tip. I see the format is corrected via clang-format. Would this be enough or should I run clang-format with a specific option?

Reply:
The .clang-format file we have in the tree *should* take care of all the options. I'd always recommend checking the diff after running the script, it can make mistakes.